api/geth_backend.go

```diff
@@ -760,25 +759,74 @@ func (b *GethStatusBackend) ChangeDatabasePassword(keyUID string, password strin
 	}
 
 	if keyDir != "" {
-		err := b.accountManager.ReEncryptKeyStoreDir(keyDir, password, newPassword)
+		err := b.accountManager.ReEncryptKeyStoreDir(keyDir, currentPassword, newPassword)
 		if err != nil {
 			return fmt.Errorf("ReEncryptKeyStoreDir error: %v", err)
 		}
 	}
+	return nil
+}
 
-	kdfIterations, err := b.multiaccountsDB.GetAccountKDFIterationsNumber(keyUID)
+func (b *GethStatusBackend) ChangeDatabasePassword(keyUID string, password string, newPassword string) error {
+	dbPath := filepath.Join(b.rootDataDir, fmt.Sprintf("%s-v4.db", keyUID))
+
+	account, err := b.multiaccountsDB.GetAccount(keyUID)
 	if err != nil {
 		return err
 	}
-	err = appdatabase.ChangeDatabasePassword(dbPath, password, kdfIterations, newPassword)
+
+	file, err := os.CreateTemp("", "*-v4.db")
 	if err != nil {
-		if config != nil {
-			keyDir := config.KeyStoreDir
-			// couldn't change db password so undo keystore changes to mainitain consistency
-			_ = b.accountManager.ReEncryptKeyStoreDir(keyDir, newPassword, password)
+		return err
+	}
+
+	newDBPath := file.Name()
+	defer func() {
+		_ = file.Close()
+		_ = os.Remove(newDBPath)
+		_ = os.Remove(newDBPath + "-wal")
+		_ = os.Remove(newDBPath + "-shm")
+		_ = os.Remove(newDBPath + "-journal")
+	}()
+
+	// Exporting database to a temporary file with a new password
+	err = appdatabase.ExportDB(dbPath, password, account.KDFIterations, newDBPath, newPassword)
+	if err != nil {
+		return err
+	}
+
+	err = b.reEncryptKeyStoreDir(password, newPassword)
+	if err != nil {
+		return err
+	}
+
+	// Replacing the old database with the new one requires closing all connections to the database
+	// This is done by stopping the node and restarting it with the new DB
+	appDBPath, _ := appdatabase.GetDBFilename(b.appDB)
+	changeCurrentAccountPassword := appDBPath == dbPath
```

Comment:
@alexjba sorry for the late check, but why do we need `changeCurrentAccountPassword` here? I mean, we're here only when we actually want to change the pass? 

Also if your intention was, if in case there are more users to let all but the logged in user to change the pass and in case of logged in user to log him out first and then change the pass, that's not needed, cause the only way yo change the pass (reencrypt the db) is to be logged in.

Reply:
> I mean, we're here only when we actually want to change the pass?

Yes, that's true.

I was thinking it might not be needed with the current flows we have. But since the API is not clear on this, I didn't want to make an assumption here.

My intention was to allow password change without requiring login. Looking at the API, it can receive a keyUID, old password, new password. This means to me the  password could be changed for any keyUID as long as the old password is correct. And I could also see some reasoning behind this. Eg. Changing password from logins screen or Deep links. 

Now since you've challenged me to think more about it, I see there's a bug here. Changing the password won't work if the user is not logged-in because the keystore dir re-encryption will probably fail as it's using `b.accountManager.Keydir`